plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/AppEngineLibraries.java

```diff
@@ -16,29 +16,44 @@
 
 package com.google.cloud.tools.eclipse.appengine.libraries;
 
-import java.util.Arrays;
-import java.util.Collections;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.RegistryFactory;
 
 import com.google.cloud.tools.eclipse.appengine.libraries.model.Library;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFactory;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFactoryException;
 
 public class AppEngineLibraries {
 
-  // TODO obtain libraries from extension registry
-  // https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/819
+  private static final Logger logger = Logger.getLogger(AppEngineLibraries.class.getName());
+  
+  // TODO obtain libraries by group name
+  // TODO cache library list
+  /**
+   * @return the three core appengine libraries
+   */
   public static List<Library> getAvailableLibraries() {
```

Comment:
yes, we may want to store this in a field so we don't constantly reparse. Static initialization order can be problematic, so I'd rather do this as lazy initialization on first call, which is also problematic but at least I understand some of the pitfalls there. I'm not as confident about when Eclipse actually makes the configuration elements available to us.

Refactoring, certainly. My thinking was that most of the other code was going to end up calling into this class and asking it to provide the libraries they need.